Storage/src/Connection/Rest.php

```diff
@@ -599,4 +640,96 @@ protected function supportsBuiltinCrc32c()
     }
+
+    /**
+     * Add the required retry function and send the request.
+     *
+     * @param string $resource resource name, eg: buckets.
+     * @param string $method method name, eg: get
+     * @param array $options [optional] Options used to build out the request.
+     * @param array $whitelisted [optional]
+     */
+    public function send($resource, $method, array $options = [], $whitelisted = false)
+    {
+        $retryMap = [
+            'projects.resources.serviceAccount' => 'serviceaccount',
+            'projects.resources.hmacKeys' => 'hmacKey',
+            'bucketAccessControls' => 'bucket_acl',
+            'defaultObjectAccessControls' => 'default_object_acl',
+            'objectAccessControls' => 'object_acl'
+        ];
+        $retryResource = isset($retryMap[$resource]) ? $retryMap[$resource] : $resource;
+        $options['restRetryFunction'] = $this->restRetryFunction ?? $this->getRestRetryFunction(
+            $retryResource,
+            $method,
+            $options
+        );
+
+        $options = $this->addRetryHeaderLogic($options);
+
+        return $this->traitSend($resource, $method, $options);
+    }
+
+    /**
+     * Adds the retry headers to $args which amends retry hash and attempt
+     * count to the required header.
+     * @param array $args
+     * @return array
+     */
+    private function addRetryHeaderLogic(array $args)
+    {
+        $invocationId = Uuid::uuid4()->toString();
+        $args['retryHeaders'] = self::getRetryHeaders($invocationId, 1);
+
+        $currentListener = $args['restRetryListener'] ?? null;
+
+        // Adding callback logic to update headers while retrying
+        $args['restRetryListener'] = function (
+            \Exception $e,
+            $retryAttempt,
+            $arguments
+        ) use (
+            $invocationId,
+            $currentListener
+        ) {
+            $changes = self::getRetryHeaders($invocationId, $retryAttempt + 1);
+            $request = $arguments[0];
+            $headerLine = $request->getHeaderLine(AgentHeader::AGENT_HEADER_KEY);
+
+            // An associative array to contain final header values as
+            // $headerValueKey => $headerValue
+            $headerElements = [];
+
+            // Adding existing values
+            $headerLineValues = explode(' ', $headerLine);
+            foreach ($headerLineValues as $value) {
+                $key = explode('/', $value)[0];
+                $headerElements[$key] = $value;
+            }
+
+            // Adding changes with replacing value if $key already present
+            foreach ($changes as $change) {
+                $key = explode('/', $change)[0];
+                $headerElements[$key] = $change;
+            }
+            $arguments[0] = Utils::modifyRequest($request, [
+                'set_headers' => [
+                    AgentHeader::AGENT_HEADER_KEY => implode(' ', $headerElements)
+                ]
+            ]);
```

Comment:
We can just use `withHeader` here, as the purpose of the `modify_request` function is to consolidate many changes happening to a response at once, and we are only making one change:
```suggestion
            $arguments[0] = $request->withHeader(
                AgentHeader::AGENT_HEADER_KEY,
                implode(' ', $headerElements)
            );
```